File: Lesson 3 - Time Complexity/TapeEquilibrium.cpp
#include <iostream>
#include <assert.h>

#include <algorithm>
#include <limits.h>
#include <vector>

int solution(std::vector<int> &A)
{
    if (A.size() < 2)
    {
        return 0;
    }

    int64_t rightPartSum = 0;
    std::for_each(A.begin(), A.end(), [&rightPartSum](const int& n) { rightPartSum += n; });

    int64_t leftPartSum = 0;
    int64_t result = INT_MAX;
    for (size_t i = 0; i < A.size() - 1; i++)
    {
        leftPartSum += A[i];
        rightPartSum -= A[i];

        int64_t localResult = std::abs(leftPartSum - rightPartSum);
        if (localResult < result)
        {
            result = localResult;
        }
    }

    return static_cast<int>(result);
}

int main()
{
    assert(0 == solution(std::vector<int>()));
    assert(0 == solution(std::vector<int>({ 1 })));
    assert(0 == solution(std::vector<int>({ 1, 1 })));
    assert(0 == solution(std::vector<int>({ -1, 1, 0 })));
    assert(1 == solution(std::vector<int>({ 3, 1, 2, 4, 3 })));
    assert(0 == solution(std::vector<int>({ INT_MAX, INT_MAX, INT_MAX, INT_MAX })));
    assert(0 == solution(std::vector<int>({ INT_MIN, INT_MIN, INT_MIN, INT_MIN })));
    assert(INT_MAX == solution(std::vector<int>({ INT_MAX, INT_MAX, INT_MAX })));
}

File: Lesson 3 - Time Complexity/TapeEquilibrium.py
from sys import maxint

def solution(A):
    if (len(A) < 2):
        return 0

    rightPartSum = sum(A)

    leftPartSum = 0
    result = maxint
    for i in range(0, len(A) - 1):
        leftPartSum += A[i]
        rightPartSum -= A[i]

        localResult = abs(leftPartSum - rightPartSum)
        if localResult < result:
            result = localResult

    return result


assert 0 == solution([])
assert 0 == solution([1])
assert 0 == solution([1, 1])
assert 0 == solution([-1, 1, 0])
assert 1 == solution([3, 1, 2, 4, 3])
assert 0 == solution([maxint, maxint, maxint, maxint])
assert 1 == solution([-maxint - 1, -maxint - 1, -maxint - 1, -maxint - 1])
assert maxint == solution([maxint, maxint, maxint])
